tests/vm_test.go

```diff
@@ -1326,61 +1323,59 @@ var _ = Describe("[rfe_id:1177][crit:medium][vendor:cnv-qe@redhat.com][level:com
 					By("Waiting for VM to be ready")
 					Eventually(ThisVM(vm), 360*time.Second, 1*time.Second).Should(BeReady())
 				})
-				DescribeTable("with a failing VMI and the kubevirt.io/keep-launcher-alive-after-failure annotation", func(keepLauncher string) {
-					// The estimated execution time of one test is 400 seconds.
-					By("Creating a Kernel Boot VMI with a mismatched disk")
-					vmi := utils.GetVMIKernelBootWithRandName()
-					vmi.Spec.Domain.Firmware.KernelBoot.Container.Image = cd.ContainerDiskFor(cd.ContainerDiskCirros)
 
-					By("Creating a VM with RunStrategyManual")
-					vm := libvmi.NewVirtualMachine(vmi)
+				DescribeTable("with a failing VMI and the kubevirt.io/keep-launcher-alive-after-failure annotation", func(keepLauncher string) {
+					By("creating a Running VM")
+					vm := libvmi.NewVirtualMachine(libvmifact.NewCirros(
+						libvmi.WithInterface(libvmi.InterfaceDeviceWithMasqueradeBinding()),
+						libvmi.WithNetwork(v1.DefaultPodNetwork()),
+						libvmi.WithAnnotation(v1.KeepLauncherAfterFailureAnnotation, keepLauncher),
+						libvmi.WithAnnotation(v1.FuncTestLauncherFailFastAnnotation, ""),
+					), libvmi.WithRunStrategy(v1.RunStrategyOnce))
 					vm.Spec.Running = nil
-					vm.Spec.RunStrategy = &runStrategyManual
 
-					By("Annotate the VM with regard for leaving launcher pod after qemu exit")
-					vm.Spec.Template.ObjectMeta.Annotations = map[string]string{
-						v1.KeepLauncherAfterFailureAnnotation: keepLauncher,
-					}
 					vm, err := virtClient.VirtualMachine(testsuite.GetTestNamespace(vm)).Create(context.Background(), vm, k8smetav1.CreateOptions{})
 					Expect(err).ToNot(HaveOccurred())
 
-					By("Starting the VMI with virtctl")
-					startCommand := clientcmd.NewRepeatableVirtctlCommand(virtctl.COMMAND_START, "--namespace", vm.Namespace, vm.Name)
-					Expect(startCommand()).To(Succeed())
-
-					By("Waiting for VM to be in Starting status")
-					Eventually(ThisVM(vm), 160*time.Second, 1*time.Second).Should(HavePrintableStatus(v1.VirtualMachineStatusStarting))
+					By("waiting for virt-launcher")
+					var vmi *v1.VirtualMachineInstance
+					Eventually(func() bool {
+						vmi, err = virtClient.VirtualMachineInstance(vm.Namespace).Get(context.Background(), vm.Name, k8smetav1.GetOptions{})
+						if errors.IsNotFound(err) {
+							return false
+						}
+						Expect(err).ToNot(HaveOccurred())
+						_, err = libpod.GetPodByVirtualMachineInstance(vmi, vm.Namespace)
+						if errors.IsNotFound(err) {
+							return false
+						}
+						Expect(err).ToNot(HaveOccurred())
+						return true
+					}, 60*time.Second, 5*time.Second).Should(BeTrue(), "virt-launcher should be created eventually")
 
-					By("Waiting for VMI to fail")
-					Eventually(ThisVMIWith(vm.Namespace, vm.Name), 480*time.Second, 1*time.Second).Should(BeInPhase(v1.Failed))
+					getComputeContainerStateRunning := func() *k8sv1.ContainerStateRunning {
```

Comment:
can we return also error here?

Reply:
Thanks, but I prefer to keep it as it is. I think it is clear enough because the function does exactly as described in its name: it returns the Compute Container State Running and not an error. I don't think we should return an error if it doesn't exist because this is expected in one of the DescribeTable entries.

Reply:
@Barakmor1 We want to avoid assertions inside helpers, if possible. We prefer returning an error and assertion for no error where calling the function.

Reply:
@nunnatsa There are many cases where we use assertions inside `Eventually`/ `Consistently` scopes please see: https://github.com/kubevirt/kubevirt/pull/11848#discussion_r1590296825

Reply:
1. This is not `Eventually` nor `Consistently`.
2. We want to write new code with higher standards.

There are disadvantages of using gomega assertions within helper function. First, that make the function aware of where it's going to be called, and so it's bad design. But more than that: gomega's error logs is misleading in these cases. You won't get the context of the failure - because you won't be able to know where it was called from. If you're going to use offset (`ExpectWithOffset`), then you loose the exact error if you have several assertions. 

The golang way is to return error. Only use assertions in the test itself.

Reply:
still. you can do

```go
getComputeContainerStateRunning := func() (*k8sv1.ContainerStateRunning, error) { ... }
```

And then:
```go
Eventually(func(g Gomega){
    ...
    err, running := getComputeContainerStateRunning()
    g.Expect(err).ToNot(HaveOccurred())
    ...
}).Should(Succeed())
```

Reply:
Again, not complaining about lambda, but about asserting within helper function.

Reply:
ok fair enough i will update this soon